Port NativeAOT exception handling to CoreCLR

janvorli opened the pull request:
This change ports NativeAOT exception handling to CoreCLR, resulting in 3.5..4 times speedup in exception handling. Due to time constraints and various complexities, thread abort and debugger support is not completed yet, so this change is enabled only when `DOTNET_EnableNewExceptionHandling` env variable is set. By default, the old way is used.
This change supports all OSes and targets we support except of x86 Windows. That may be doable too in the future, but the difference in exception handling on x86 Windows adds quite a lot of complexity into the picture.

Notes for the PR:
* I have left the `ExceptionHandling.cs` and `StackFrameIterator.cs` in the nativeaot folder to simplify the review. I can move it to some common location after the change is reviewed. Also it was not clear to me where that should be, so advise would be welcome here.
* Naming of the native helpers like `RhpCallCatchFunclet` was left the same as in the NativeAOT for now.
* There are still some little things I'd like to eventually clean up, like `ExInfo` encapsulation and possibly moving `REGDISPLAY` and `CONTEXT` it uses into the `ExInfo` itself or moving debug members of `StackFrameIterator` and `REGDISPLAY` to the end of those structures so that the `AsmOffsets.cs` can be simplified. It also may be possible to unify the exception handling callback that's used for ObjectiveC to use the managed version. I've tried and there were some ugly complications, so I've left it separated.
* There are two bug fixes for bugs unrelated to this PR and a removal of unused parameter in existing code that could be made as separate PRs before this PR.
  * `ProfilerEnter` and `ProfilerLeave` for the case of `UnmanagedCallersOnly` method were being called in preemptive mode.
  * NativeAOT code for rethrowing exception was incorrectly calling `DispatchEx` with last argument set to `activeExInfo._idxCurClause` to start at the last clause processed when the rethrown exception was originally thrown instead of starting from the first one again. I have a accidentally came with a simple test that discovered this bug and causes failures in  the original NativeAOT too.
* Changes in  the stackwalk.cpp add support for
  * Usage of `ExInfo` instead of `ExceptionTracker`
  * Handling of case when GC runs while finally funclet is on the stack and then again when the code is back in the new exception handling code in managed code before other finally or catch funclet is called. The NativeAOT solves that by disabling GC for the 2nd pass of EH, for this change it would not be reasonable.
  * Handling the GC reporting when funclet is found while walking the stack. It needs to scan frames of the managed code that handles the exception too,  since it contains live references. The old EH way doesn't have this case.
* I needed to add `GCFrame::Remove` method that can remove the `GCFrame` from any location in the chain. There is a managed runtime method that calls `GCReporting::Unregister` that was popping it with my changes out of order due to the exception handling code being managed.

janvorli committed:
Port NativeAOT exception handling to CoreCLR

This change ports NativeAOT exception handling to CoreCLR, resulting in
3.5..4 times speedup in exception handling. Due to time constraints and
various complexities, thread abort and debugger support is not
completed yet, so this change is enabled only when
`DOTNET_EnableNewExceptionHandling` env variable is set. By default,
the old way is used.
This change supports all OSes and targets we support except of x86
Windows. That may be doable too in the future, but the difference in
exception handling on x86 Windows adds quite a lot of complexity into
the picture.

Notes for the PR:
* I have left the `ExceptionHandling.cs` and `StackFrameIterator.cs` in
  the nativeaot folder to simplify the review. I can move it to some
  common location after the change is reviewed. Also it was not clear to
  me where that should be, so advise would be welcome here.
* Naming of the native helpers like `RhpCallCatchFunclet` was left the
  same as in the NativeAOT for now.
* There are still some little things I'd like to eventually clean up,
  like `ExInfo` encapsulation and possibly moving `REGDISPLAY` and
  `CONTEXT` it uses into the `ExInfo` itself or moving debug members of
  `StackFrameIterator` and `REGDISPLAY` to the end of those structures
  so that the `AsmOffsets.cs` can be simplified. It also may be possible
  to unify the exception handling callback that's used for ObjectiveC to
  use the managed version. I've tried and there were some ugly
  complications, so I've left it separated.
* There are two bug fixes for bugs unrelated to this PR and a removal of
  unused parameter in existing code that could be made as separate PRs
  before this PR.
  * `ProfilerEnter` and `ProfilerLeave` for the case of
    `UnmanagedCallersOnly` method were being called in preemptive mode.
  * NativeAOT code for rethrowing exception was incorrectly calling
    `DispatchEx` with last argument set to `activeExInfo._idxCurClause`
    to start at the last clause processed when the rethrown exception
    was originally thrown instead of starting from the first one again.
    I have a accidentally came with a simple test that discovered this
    bug and causes failures in  the original NativeAOT too.
* Changes in  the stackwalk.cpp add support for
  * Usage of `ExInfo` instead of `ExceptionTracker`
  * Handling of case when GC runs while finally funclet is on the stack
    and then again when the code is back in the new exception handling
    code in managed code before other finally or catch funclet is
    called. The NativeAOT solves that by disabling GC for the 2nd pass
    of EH, for this change it would not be reasonable.
  * Handling the GC reporting when funclet is found while walking the
    stack. It needs to scan frames of the managed code that handles the
    exception too,  since it contains live references. The old EH way
    doesn't have this case.
* I needed to add `GCFrame::Remove` method that can remove the `GCFrame`
  from any location in the chain. There is a managed runtime method that
  calls `GCReporting::Unregister` that was popping it with my changes
  out of order due to the exception handling code being managed.

Fix context initialization after rebase

jkotas commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -10754,8 +10754,11 @@ void __stdcall ProfilerUnmanagedToManagedTransitionMD(MethodDesc *pMD,
 
-HCIMPL2(EXTERN_C void, ProfileEnter, UINT_PTR clientData, void * platformSpecificHandle)
+HCIMPL2_RAW(EXTERN_C void, ProfileEnter, UINT_PTR clientData, void * platformSpecificHandle)
+GCX_COOP();
```
The GC mode for profiler enter/leave callbacks was discussed in https://github.com/dotnet/runtime/issues/68760

It would be best to submit this change as separate PR. I do not think that this fix is complete. The thread is not guaranteed to be setup when this callback is called.

AaronRobinsonMSFT replied:
Additionally, we should run that through the diagnostics tests quickly and confirm we aren't triggering any asserts/violations. @hoyosjs or I can help with that.

jkotas replied:
This was split into https://github.com/dotnet/runtime/pull/88134. The change should be reverted in this PR.

jkotas commented on src/coreclr/vm/frames.h:
```diff
@@ -2868,8 +2870,13 @@ class InlinedCallFrame : public Frame
     // m_Datum contains MethodDesc ptr or
-    // - on AMD64: CALLI target address (if lowest bit is set)
-    // - on X86: argument stack size (if value is <64k)
+    // - on 64 bit host: CALLI target address (if lowest bit is set)
```
Does https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/amd64/PInvokeStubs.asm#L33-L38 need to be adjusted for the stolen bits?

janvorli replied:
It doesn't need to be updated. When the lowest bit is set, we never fiddle with the newly added bits.

janvorli replied:
I'll update the comment to make it clearer.

jkotas commented on src/coreclr/debug/daccess/dacdbiimplstackwalk.cpp:
```diff
@@ -433,6 +449,20 @@ ULONG32 DacDbiInterfaceImpl::GetCountOfInternalFrames(VMPTR_Thread vmThread)
+        {
+            // Skip new exception handling helpers
+            InlinedCallFrame *pInlinedCallFrame = (InlinedCallFrame *)pFrame;
+            PTR_NDirectMethodDesc pMD = pInlinedCallFrame->m_Datum;
```
`m_Datum` is not always initialized. For example, it is not initialized by `JIT_PInvokeBegin/End` helpers. This check does not look reliable.

jkotas replied:
Right, it means that this check won't kick in as expected when we use `JIT_PInvokeBegin` for the PInvoke transition.

jkotas replied:
Ah, I see what you mean. Ignore my comment.

jkotas commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -2315,9 +2315,9 @@ void CodeGen::genReportEH()
+        if (XTnum > 0)
         {
-            // For NativeAOT, CORINFO_EH_CLAUSE_SAMETRY flag means that the current clause covers same
+            // For NativeAOT and the new coreclr EH, CORINFO_EH_CLAUSE_SAMETRY flag means that the current clause covers same
```
It may be a good idea to submit the JIT change as separate PR to make it easier to review and verify that it does not regress the existing EH.

AustinWise commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -664,10 +787,12 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
-#endif // FEATURE_OBJCMARSHAL
 
-            if (pCatchHandler == null && pReversePInvokePropagationCallback == IntPtr.Zero)
+            if (pCatchHandler == null && pReversePInvokePropagationCallback == IntPtr.Zero && !unwoundReversePInvoke)
```
```suggestion
            if (pCatchHandler == null && pReversePInvokePropagationCallback == IntPtr.Zero
#if !NATIVEAOT
                && !unwoundReversePInvoke
#endif
                )
```

Currently NativeAot fails fast here if there is no `pCatchHandler` or `pReversePInvokePropagationCallback`. If I'm reading the current state of  this PR correctly, if this change would cause the NativeAot to pass the null `pCatchHandler` to `RhpCallCatchFunclet` where it would try to jump to the null pointer. I'm pretty sure this would prevent NativeAot from printing the nice managed stack trace on an unhandled exception.

janvorli replied:
@AustinWise thank you for spotting this! You are right that the new part of the condition is meant only for the non-native aot case. I was just investigating why the nativeaot exception handling test was failing in this PR, so you've saved me some valuable time!

danmoseley commented:
How do the perf improvements compare between OS?

In a typical workload, would we expect this to only be noticeable in "exception storms" (e.g. due loss of connectivity)?

jkotas commented:
> How do the perf improvements compare between OS?

https://github.com/dotnet/runtime/issues/77568#issuecomment-1453934050

janvorli commented:
I have found (by running a separate testing PR (#88113) just for the NativeAOT fix for rethrowing, that that fix is not correct in all cases. So clearly the last argument of the DispatchEx needs to be sometimes `activeExInfo._idxCurClause` and sometimes `MaxTryRegionIdx`. After debugging the issue that lead me to change this, I was convinced that it should always be `MaxTryRegionIdx`, which means to start at the first clause. So I'll need to debug the case failing in the PR to see why it is not the case here. 
For the current PR, I am going to revert this change. The failure I have seen (exception going unhandled) was not happening in any of the tests we have, I was just lucky to create a test myself that was failing without the change for both NativeAOT and the new EH.

janvorli commented:
Fortunately, I have found what was causing the NativeAOT test suite to fail with my rethrow fix. It was just a problem of stack trace, the actual fix is correct.

AaronRobinsonMSFT commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -4445,7 +4445,7 @@ FCIMPL1(void, GCReporting::Unregister, GCFrame* frame)
     // Destroy the GCFrame.
     _ASSERTE(frame != NULL);
-    frame->Pop();
+    frame->Remove();
```
The `Push()`/`Pop()` names were introduced intentionally to provide naming symmetry. Is there a reason the `Pop()` can't be reimplemented as `Remove()`? If the `Remove()` is really so fundamentally different and we need the existing `Pop()`, I would instead call it `PopAll()` or something that aligns with the `Push()` above in `GCReporting::Register()`.

janvorli replied:
Please note that this is not replacing `Pop` for the GC frame chain in general. This `GCReporting::Unregister` is a special case when with the new exception handling enabled, there can be GC frames pushed after the one pushed by the `GCReporting::Register`. I'd prefer keeping the `Push` / `Pop` for all the other cases, as it verifies that we are not popping out of order. `PopAll` sounds like we would be popping all frames until the passed in frame. I see the operations of `Push` /  `Pop` / `Remove` as operations on the underlying linked list

AaronRobinsonMSFT replied:
> `Remove` as operations on the underlying linked list

Ah, so this really is a remove this particular frame from anywhere in the call stack/frame chain. I retract my comment. This makes sense.

AaronRobinsonMSFT commented on src/coreclr/vm/exceptionhandling.h:
```diff
@@ -39,6 +44,16 @@ enum EHFuncletType
+
+enum class InlinedCallFrameMarker
+{
+    ExceptionHandlingHelper = 2,
```
If the 1 bit is important we should document that here or point to where it is defined in another enum.

janvorli replied:
Preexisting code is using the bit 0 for unrelated purposes, it is described in comment on the m_Datum field. I'll add a note here.

AaronRobinsonMSFT commented on src/coreclr/vm/exceptmacros.h:
```diff
@@ -308,8 +313,42 @@ VOID DECLSPEC_NORETURN DispatchManagedException(PAL_SEHException& ex, bool isHar
+#ifdef FEATURE_EH_FUNCLETS
+#define INSTALL_MANAGED_EXCEPTION_DISPATCHER                                                \
+    {                                                                                       \
+        auto body = [&]()                                                                   \
```
Capturing all locals by reference can cause confusion in many cases. Is this really needed? I appreciate what we are trying to do here and make it generic, but it is very dangerous to "capture everything". This mixing of macro soup _and_ modern C++ capture semantics is likely going to make this very complicated to reason about in crashes.

janvorli replied:
I don't see how we could do it differently. This macro is used at several hundreds of places, so specifying explicit capture list would be a nightmare.
If you have some other idea how to make a piece of code between these macros conditionally run under try / catch or out of that, I'd love to use it instead. I wasn't able to find any other way (other than capturing it in a macro and effectively duplicating the code twice).

jkotas replied:
Is there a perf impact from "capture everything" approach?

janvorli replied:
I was going to run performance benchmarks to see.

davidwrighton replied:
Is there a reason you couldn't use the try/catch pattern even with the old style eh? While the trick with lambda's is wonderfully clever (and I used in the green threads experiment last year), it carries with it a fairly expensive abstraction penalty that I suspect an extra try/catch does not.

janvorli replied:
I have actually tried that initially, it worked for most of the cases, but there was a problem with `NullReferenceException`. This exception is propagated as SEH exception, so when C++ catch(...) catches that and then rethrows it, it gets converted into access violation (because the internal C++ exception handling code just looks at the exception code and doesn't take the failure address into account). And I couldn't put in an extra SEH __try / __except, because these cannot be combined with C++ EH in the same function (and creating an extra function for the body gets us back to the lambda).

janvorli replied:
I have ran some dotnet/performance microbenchmarks to see the performance impact of the added lambda. First, I was seeing about 15% perf hit in some exception handling tests, but I was able to mitigate it completely by moving the new eh branch stuff to a separate function. 
However, I have realized today that I actually don't need any code here even for the new exception handling. It was not the case until I've added transitioning to the new exception handling in the personality routine, which was done quite late in the development process. Now that the transition is there, it is no longer necessary to catch the exception at places where the `INSTALL_MANAGED_EXCEPTION_DISPATCHER` is used, the personality routine will do the job for us.

janvorli replied:
Removed the INSTALL_MANAGED_EXCEPTION_DISPATCHER  / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER changes for Windows completely, so there is no more lambda / perf impact here.

AaronRobinsonMSFT commented on src/coreclr/vm/exinfo.cpp:
```diff
@@ -300,4 +300,32 @@ void ExInfo::SetExceptionCode(const EXCEPTION_RECORD *pCER)
+#ifndef DACCESS_COMPILE
+ExInfo::ExInfo(Thread *pThread, CONTEXT *pCtx, REGDISPLAY *pRD, ExKind exceptionKind)
+{
+    m_pPrevExInfo = pThread->GetExceptionState()->GetCurrentExInfo();
```
Is it possible to do this initialization in the initializer list? I'm trying to make the codebase more canonical C++ and I don't see anything that should make this hard. In fact, `m_ClauseForCatch{}` is far clearer than performing the `memset()` and avoids all sorts of UB.

janvorli replied:
I can change it to that.

janvorli commented on src/coreclr/vm/amd64/cgencpu.h:
```diff
@@ -353,13 +353,15 @@ inline DWORD64 GetSSP(const CONTEXT * context)
     }
-
+#else
+    // TODO: implement for Unix
```
It is intentional that these methods return 0 when there is no shadow stack enabled or supported. The code path should be executed on Unix amd64 too, the caller relies on the returned 0 to detect it should not try to perform shadow stack massaging. The TODO is meant as "TODO when we enable Intel CET on Unix". I can modify the comment to say that explicitly or remove the `#else` block for now. I have put that in as a marker for the future when Linux will start supporting  Intel CET.

AaronRobinsonMSFT replied:
Ah. Looking more closely at the Windows impl that makes sense. Adding "when we enable Intel CET on Unix" would help with that. Thanks.

janvorli commented on src/coreclr/vm/frames.cpp:
```diff
@@ -994,6 +994,49 @@ void GCFrame::Push(Thread* pThread)
     pThread->SetGCFrame(this);
 }
 
+void GCFrame::Remove()
```
Ok, makes sense. Although I doubt this is honored in the runtime in general.

AaronRobinsonMSFT replied:
It isn't. One of these days I am hoping it will be.

jkotas commented on src/coreclr/vm/fcall.h:
```diff
@@ -582,8 +582,9 @@ LPVOID __FCThrowArgument(LPVOID me, enum RuntimeExceptionKind reKind, LPCWSTR ar
         HELPER_METHOD_FRAME_BEGIN_EX_BODY(ret, helperFrame, gcpoll, allowGC)    \
             /* <TODO>TODO TURN THIS ON!!!   </TODO> */                    \
             /* gcpoll; */                                                       \
+            if (g_isNewExceptionHandlingEnabled) __helperframe.Push();         \
```
This has risk of introducing regressions in assorted set of micro-benchmarks.

janvorli replied:
I have ran a subset of the microbenchmarks and this was not causing any visible effect.

davidwrighton commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3374,25 +3376,30 @@ BOOL StackTraceInfo::AppendElement(BOOL bAllowAllocMem, UINT_PTR currentIP, UINT
-    if (pThread && (currentIP != 0))
+#ifdef FEATURE_EH_FUNCLETS
+    if (!g_isNewExceptionHandlingEnabled)
+#endif // FEATURE_EH_FUNCLETS
```
Could you describe the impact of your change on Watson bucketing? It appears that the ExceptionTracker was involved in the bucketing system, but the ExInfo logic that you've integrated is not. In addition, my reading of this is that we may be calling the "FallbackFailFast" function when there is an unhandled exception, and I don't know how that interacts with the Watson behavior. I'm particularly concerned about triage dumps and ensuring that they continue to contain the correct details about what exception was unhandled, and where it was thrown from.

janvorli replied:
The watson stuff is not completed yet in this first version. I've mentioned that debugging stuff is not fully working in the PR description and the watson is similar case. I will work on these in the next step either for some .NET 8 update or .NET 9. These are the main reasons why it is not enabled by default.

davidwrighton commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -75,18 +99,30 @@ private struct EHEnum
         // This is a fail-fast function used by the runtime as a last resort that will terminate the process with
         // as little effort as possible. No guarantee is made about the semantics of this fail-fast.
-        internal static void FallbackFailFast(RhFailFastReason reason, object unhandledException)
+#if !NATIVEAOT
```
Since this appears to be used actively in the CoreCLR case, we should describe the semantics here better, or at least describe the semantics we are relying on.

janvorli replied:
With this change, I wanted to make it easier to review the changes in this file by not refactoring / renaming stuff / moving the file. I'd like to tackle it in a next step.

jkotas reviewed (commented):
Comments from detailed review of 1st half of the changes.

AustinWise commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -732,9 +865,11 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
 #if FEATURE_OBJCMARSHAL
             if (pReversePInvokePropagationCallback != IntPtr.Zero)
             {
+#if NATIVEAOT
```
For CoreCLR, control is transferred to the propagation callback `RhpCallCatchFunclet`:

https://github.com/janvorli/runtime/blob/port-nativeaot-eh-to-coreclr-final-2/src/coreclr/vm/exceptionhandling.cpp#L7683-L7690

janvorli replied:
Right, I have implemented that before the managed EH had the Objective C interop implemented. When I've rebased my changes on top of main before creating the PR, I've found about it and tried to move to the managed implementation. However, it got somehow ugly. I would need to call an extra native helper to get things that we get in `ObjCMarshalNative::GetPropagatingExceptionCallback` in the current implementation like:
```
        ILStubResolver* resolver = dynamicMethod->GetILStubResolver();
        MethodDesc* methodMaybe = resolver->GetStubTargetMethodDesc();
```
and I think also
```
        gc.methodRef = method->GetStubMethodInfo();
```
as it doesn't seem there is any way to get it via existing stuff.

IIRC it was also complicating the `RhpCallCatchFunclet` for the case of Objective C, but I've forgotten the details.

I am not saying we should keep the implementation split for NativeAOT / CoreCLR forever, but the way I've done it seemed easier for the first version and easier to reason about.

AustinWise commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -656,10 +776,16 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
                         catchingTryRegionIdx = MaxTryRegionIdx;
```
This if statement can be removed, since `handlingFrameSP` and `catchingTryRegionIdx` are being set to the same values below.

```c#
if (pReversePInvokePropagationCallback != IntPtr.Zero)
{
    // Tell the second pass to unwind to this frame.
    handlingFrameSP = frameIter.SP;
    catchingTryRegionIdx = MaxTryRegionIdx;
}
```

janvorli replied:
This is actually a bug I've introduced during final cleanup. They should be set below only for the non-NATIVEAOT case.

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -5322,7 +5387,40 @@ BOOL HandleHardwareException(PAL_SEHException* ex)
+            DWORD exceptionCode = ex->GetExceptionRecord()->ExceptionCode;
+            if (exceptionCode == STATUS_ACCESS_VIOLATION)
+            {
+                #define NULL_AREA_SIZE   (64 * 1024)
```
This is Windows-specific definition that is not correct for Unix. Move the definition from excep.cpp to a shared place instead.

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+    bool canAllocateMemory = !(exceptionObj.Get() == CLRException::GetPreallocatedOutOfMemoryException()) &&
+                             !(exceptionObj.Get() == CLRException::GetPreallocatedStackOverflowException());
+
+    EECodeInfo codeInfo(ip);
```
The stackwalker must have resolved IP to MethodDesc already. Is this redundant lookup leaving perf on the table?

janvorli replied:
I can extract the `MethodDesc` from the passed in `ExInfo`.

janvorli replied:
Done

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+        SetIP(pvRegDisplay->pCurrentContext, dwResumePC);
+    }
+
+    ULONG64 targetSp = GetSP(pvRegDisplay->pCurrentContext);
```
```suggestion
    UINT_PTR targetSp = GetSP(pvRegDisplay->pCurrentContext);
```
ULONG64 does not make sense on 64-bit here

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+        // use thread local storage
+        thread_local OBJECTREF oref;
+        oref = exceptionObj.Get();
+        SetIP(pvRegDisplay->pCurrentContext, (ULONG64)(void (*)(OBJECTREF))RealCOMPlusThrow);
```
```suggestion
        SetIP(pvRegDisplay->pCurrentContext, (PCODE)(void (*)(OBJECTREF))RealCOMPlusThrow);
```

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+#endif
+
+#ifdef USE_CHECKED_OBJECTREFS
+        pvRegDisplay->pCurrentContext->FIRST_ARG_REG = (size_t)&oref;
```
This makes an assumptions about calling convention used to pass `OBJECTREF`. I do not think that this will work on all platforms.

Would it be better to have a dedicated entrypoint that takes `Object*` and avoid the avoid the thread local, etc.?

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+        pEHClause->_tryEndOffset = EHClause.TryEndPC;
+        if (IsFilterHandler(&EHClause))
+        {
+            pEHClause->_filterAddress =  (BYTE*)pJitMan->GetCodeAddressForRelOffset(MethToken, EHClause.FilterOffset);
```
`GetCodeAddressForRelOffset` is an expensive call when the code is hot/cold split. Can we delay it to only once we need the address like it is done in the current exception handling?

janvorli replied:
That would mean diverging between NativeAOT and the new EH semantics of the `_filterAddress` field though.

jkotas replied:
We can update NativeAOT to do the same. The NativeAOT semantics are not necessarily the best unified semantics.

This can be `TODO-NewEH`.

janvorli replied:
Added TODO-NewEH

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+                //  * Notifify debugger and return.
+                //      In this case the normal EH proceeds and we need to reset m_sfResumeStackFrame to the sf catch handler.
+
+                // TODO: New exception handling debugger events completion
```
```suggestion
                // TODO-NewEH: New exception handling debugger events completion
```

janvorli replied:
updated

janvorli commented:
@tannergooding the tests failing are six HardwareIntrinsics tests that all fail with:
```
System.Runtime.Intrinsics.Vector256.IsHardwareAccelerated returned true but the runtime returned false
System.Runtime.Intrinsics.Vector512.IsHardwareAccelerated returned true but the runtime returned false
```
I've been hitting this about two weeks ago locally on some virtual machines (devbox) in both Windows and Linux even without any of the changes in this PR.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/InternalCalls.cs:
```diff
@@ -0,0 +1,48 @@
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime
```
```suggestion
namespace System.Runtime.ExceptionServices
```
to match the file path?

jkotas commented on src/coreclr/vm/qcallentrypoints.cpp:
```diff
@@ -320,6 +322,16 @@ static const Entry s_QCall[] =
     DllImportEntry(ObjectToComWeakRef)
 #endif
+#ifdef FEATURE_EH_FUNCLETS
+    DllImportEntry(RhpSfiInit)
```
We should avoid introducing Rh... naming convention in CoreCLR. We ported some Rh... helpers in the past, but it did not look good and it did not age well. They have been renamed since then.

I hope that we will get rid of Rh names in NAOT at some point.

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -244,13 +311,25 @@ private enum RhEHFrameType
                 isFirstFrame = false;
             }
+#else
+#pragma warning disable CS8500
```
That doesn't compile:
`error CS0306: The type 'EH.ExInfo' may not be used as a type argument`. I guess it is because of the `object _exception` member of the struct.

jkotas replied:
It is because of it is a `ref struct`.

janvorli commented on src/coreclr/vm/amd64/Context.S:
```diff
@@ -0,0 +1,70 @@
+        // exception handling, iret and ret can't be used because their shadow stack enforcement would not allow that transition,
+        // and using them would require writing to the shadow stack, which is not preferable. Instead, iret is partially
+        // simulated.
+        mov     eax, [r10 + OFFSETOF__CONTEXT__EFlags]
```
This was just a port of the Windows amd64 code, you are right that on Unix it is not needed to restore EFlags and xmm registers.

janvorli replied:
Fixed

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -867,6 +873,40 @@ ProcessCLRException(IN     PEXCEPTION_RECORD   pExceptionRecord,
+        {
+            if (pExceptionRecord->ExceptionCode == 0x80000003)
+            {
+                // Assert breakpoint
```
The breakpoint instruction fault that is fired from `_ASSERTE` was in some cases leaking into this code, leading to infinite recursion. This was actually a hot fix for my local testing. I need to check how that was happening, it seems it should have been handled in the VEH.

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -75,18 +99,30 @@ private struct EHEnum
+#if NATIVEAOT
             InternalCalls.RhpFallbackFailFast();
+#else
+            System.Diagnostics.Debugger.Break();
```
This is basically just a help I've used for debugging, it should not be there. The `Environment.FailFast` will probably need to be converted to something different, as when running under debugger, it just exits the process and the stack at which the issue occurred is not present anymore. This is another debugger related "TODO" for the follow up changes.

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -2,18 +2,39 @@
+        }
+        internal static void OnUnhandledException(object e)
+        {
+            UnhandledException?.Invoke(AppDomain.CurrentDomain, new UnhandledExceptionEventArgs(e, true));
```
There were tests failing without this being implemented here because they were not getting the notification, but it might be that it was just a point in time during the feature development and that now all unhandled exception end up going through the unmanaged code. I'll test it to see.

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -867,6 +873,40 @@ ProcessCLRException(IN     PEXCEPTION_RECORD   pExceptionRecord,
+            DispatchManagedException(oref);
+        }
+#endif // HOST_UNIX
+        EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE(E_FAIL, _T("SEH exception leaked into managed code"));
```
Makes sense

janvorli replied:
Done

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -875,7 +1030,7 @@ private static void DebugVerifyHandlingFrame(UIntPtr handlingFrameSP)
         }
 
-#if DEBUG && !INPLACE_RUNTIME
+#if DEBUG && !INPLACE_RUNTIME && NATIVEAOT
```
I am not sure I understand what the `INPLACE_RUNTIME` means. The `NATIVEAOT` seemed more obvious. What is the motivation for using the `INPLACE_RUNTIME` instead? As for simplification, this seems to be the only place where it simplifies the condition.

MichalStrehovsky replied:
INPLACE_RUNTIME means these files are compiled into System.Private.CoreLib assembly instead of into Runtime.Base assembly. On .NET Native we compiled these into Runtime.Base that had it's own copy of System.Object, etc.

If INPLACE_RUNTIME is not defined, it means we're compiling Runtime.Base at it existed in .NET Native. We don't currently have such build. We might have it in the future.

jkotas replied:
> As for simplification, this seems to be the only place where it simplifies the condition.

There are more places where it is more appropriate condition, e.g. here: https://github.com/dotnet/runtime/pull/88034/files#r1254740518

janvorli replied:
@jkotas I am still not sure I am on the same page with you here. Should every single `#if NATIVEAOT` be converted to `#if INPLACE_RUNTIME` in the ExceptionHandling.cs? According to @MichalStrehovsky said, it is possible that eventually the NativeAOT will also support mode with `INPLACE_RUNTIME` defined. But many of the changes I've made here won't be valid for NativeAOT.

MichalStrehovsky replied:
I think what Jan meant is to add INPLACE_RUNTIME to the list of defined symbols when compiling CoreCLR's corelib. It kind of makes sense since this code is included from the real CoreLib and that's the meaning of INPLACE_RUNTIME on NativeAOT side.

jkotas replied:
Yes, it is what I meant.

It does not mean converting every single `#if NATIVEAOT` to `#if INPLACE_RUNTIME`. `#if INPLACE_RUNTIME` should be only used to ifdef out base runtime abstractions like `RhpGetClasslibFunctionFromEEType`.

janvorli replied:
Got it. However, after making the change, I have found that there is a project `src/coreclr/nativeaot/Test.CoreLib/src/Test.CoreLib.csproj` which now fails to build because it also defines `INPLACE_RUNTIME`. What is that meant for? From the errors, it seems I'll need to add the `AppContext.OnFirstChanceException` / `OnUnhandledException`. It also cannot find couple of exception types in the `GetClasslibException` function, namely `AccessViolationException`, `AmbiguousImplementationException`, `DataMisalignedException` and `EntryPointNotFoundException` when compiling this project.

jkotas replied:
Ok, feel free to ignore this suggestion for this PR.

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7165,5 +7437,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+        else
+        {
+            size_t CallDescrWorkerInternalReturnAddress = (size_t)CallDescrWorkerInternal + CallDescrWorkerInternalReturnAddressOffset;
+            if (GetIP(pThis->m_crawl.GetRegisterSet()->pCallerContext) == CallDescrWorkerInternalReturnAddress)
```
I have found a related comment here:
https://github.com/dotnet/runtime/blob/4a8a0eea8b24a97ed9e37e86bcfc93568eee8fb8/src/coreclr/vm/callhelpers.cpp#L43-L48

It seems (based on the `CallDescrWorkerUnwindFrameChainHandler` personality routine usage) that besides the `FastCallFinalizeWorker` only the `COMToCLRDispatchHelperWithStack` may need to be taken care of. The `ArrayInitializeWorker` mentioned in the comment doesn't exist.

janvorli replied:
Actually, the `COMToCLRDispatchHelperWithStack` seems to be a dead code.

jkotas commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -897,16 +1052,42 @@ private static void AssertNotRuntimeObject(MethodTable* pClauseType)
+                    mt = mt->ParentMethodTable;
+                }
+
+                if (exception is RuntimeWrappedException ex)
```
> figure out if an exception should be unwrapped using managed code

The exception wrapping/unwrapping is controlled per-assembly by a custom attribute. In theory, you can parse assembly custom attributes using reflection, but I do not think it is a good idea to depend on reflection in core exception handling.

janvorli replied:
Fixed

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -388,12 +493,19 @@ private enum HwExceptionCode : uint
+            [FieldOffset(AsmOffsets.OFFSETOF__PAL_LIMITED_CONTEXT__ContextFlags)]
+            internal uint ContextFlags;
+            [FieldOffset(AsmOffsets.OFFSETOF__PAL_LIMITED_CONTEXT__SP)]
+            internal IntPtr SP;
```
Done

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -2,18 +2,39 @@
+#if !NATIVEAOT
+namespace System
+{
+    public static partial class AppContext
```
Done

janvorli commented on src/coreclr/debug/daccess/dacdbiimplstackwalk.cpp:
```diff
@@ -261,6 +261,22 @@ BOOL DacDbiInterfaceImpl::UnwindStackWalkFrame(StackWalkHandle pSFIHandle)
                 continue;
             }
+#ifdef FEATURE_EH_FUNCLETS
+            else if (g_isNewExceptionHandlingEnabled && pIter->GetFrameState() == StackFrameIterator::SFITER_FRAMELESS_METHOD)
```
Fixed

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -4214,6 +4216,59 @@ HCIMPLEND
 /*************************************************************/
 
+#ifdef FEATURE_EH_FUNCLETS
+HCIMPL1(void, IL_ThrowNew,  Object* obj)
```
Diagnostic tools have `IL_Throw` symbol name hard-coded in number of places, e.g. here:
https://microsoft.visualstudio.com/OS/_search?action=contents&text=IL_Throw&type=code&lp=code-Project&filters=ProjectFilters%7BOS%7DRepositoryFilters%7Bos.2020%7D&pageSize=25&includeFacets=false&result=DefaultCollection/OS/os.2020/GBofficial/main//sdktools/debuggers/exts/badev/extdll/uanalysis.cpp

This rename is going to require chasing down all these places and fixing them up, I am not sure that it is worth it.

janvorli replied:
Reverted

jkotas commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -860,6 +860,12 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\ConstrainedExecution\ReliabilityContractAttribute.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)..\..\..\coreclr\nativeaot\Runtime.Base\src\System\Runtime\ExceptionHandling.cs" Condition="'$(FeatureNativeAot)' != 'true' and '$(FeatureMono)' != 'true' and ('$(Platform)' != 'x86' or '$(TargetsWindows)' != 'true')"/>
+    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\ExceptionServices\AsmOffsets.cs" Condition="'$(FeatureNativeAot)' != 'true' and '$(FeatureMono)' != 'true' and ('$(Platform)' != 'x86' or '$(TargetsWindows)' != 'true')"/>
```
These are CoreCLR specific files. It would make more sense to have them included via https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj . it is going to make the condition simpler.

janvorli replied:
Makes sense, I wasn't sure about where they really belong.

janvorli replied:
Done

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/AppContext.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public static partial class AppContext
+    {
+        internal static void OnFirstChanceException(object e)
```
(I would be ok with having #ifdef around it if necessary. It is better than having two AppContext.cs files in different directories.)

janvorli replied:
Done

janvorli commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -2302,10 +2302,10 @@ void CodeGen::genReportEH()
         CORINFO_EH_CLAUSE_FLAGS flags = ToCORINFO_EH_CLAUSE_FLAGS(HBtab->ebdHandlerType);
 
-        if (isNativeAOT && (XTnum > 0))
+        if (XTnum > 0)
```
Done

davidwrighton commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -7159,5 +7446,839 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+                EEToProfilerExceptionInterfaceWrapper::ExceptionCatcherLeave();
+            }
+        }
+        // TODO: add the m_EHClauseInfo for profiler 
```
What is this TODO about?

davidwrighton replied:
Should it be a TODO-NewEH

janvorli replied:
Right, I still plan to add a number of TODO-NewEHs to this PR. As for this TODO, the comment was related to the fact that the `ExceptionTracker::MakeCallbacksRelatedToHandler`, which is the old EH variant of this function, calls `m_EHClauseInfo.ResetInfo()` and IIRC, profiler stuff somehow relied on more stuff from the `m_EHClauseInfo` info than the stuff I've extracted from it and use in the new EH.

davidwrighton commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -963,6 +963,14 @@ BOOL Thread::ReadyForAsyncException()
+#ifdef FEATURE_EH_FUNCLETS
+    if (g_isNewExceptionHandlingEnabled)
+    {
+        // TODO: make thread abort work for the new exception handling
```
Agreed. This needs to be marked with the TODO-NewEH notation

janvorli committed:
Remove Rhp prefix from the native helpers

janvorli committed:
Fix handling of exceptions from JIT

The `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE` in the
`EE_TO_JIT_TRANSITION` needs to rethrow an exception (if any) using native
exception handling mechanism instead of calling the new managed
exception handling, as the native exception needs to propagate through
some native code layers from there.
This change adds parameter to the
`UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE` macro to select whether
to rethrow the exception as native or to invoke the new managed
exception handling.

This problem didn't show up until I ran the coreclr tests with tiered
compilation disabled.

janvorli commented:
@jkotas I believe I have addressed all of the feedback. Can you please take a look to see if you have any other comments or if it can be merged? The CI failures are unrelated.

jkotas commented:
> The `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE` in the
`EE_TO_JIT_TRANSITION` needs to rethrow an exception (if any) using native
exception handling mechanism instead of calling the new managed
exception handling, as the native exception needs to propagate through
some native code layers from there.

Is JIT/EE interface really the only place that needs this change? I went over all `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER` occurences and it seems that some of them need this fix too. For example, this one: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/excep.cpp#L6055-L6057

janvorli commented:
> Is JIT/EE interface really the only place that needs this change? I went over all `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER` occurences and it seems that some of them need this fix too

That's a good point, I'll review the usage. The place that you've mentioned definitely needs it.

janvorli committed:
Update UNINSTALL_UNWIND_AND_CONTINUE_HANDLER usage

There were three places where the UNINSTALL_UNWIND_AND_CONTINUE_HANDLER
needed to be replaced by
UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE(true).

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -6052,7 +6063,7 @@ CreateCOMPlusExceptionObject(Thread *pThread, EXCEPTION_RECORD *pExceptionRecord
             GCPROTECT_END();
 
-            UNINSTALL_UNWIND_AND_CONTINUE_HANDLER;
+            UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE(true);
```
`NO_PROBE` looks like left-over from stackoverflow handling. It has nothing to do with what the macro does now.

Could you please rename it to something like `INSTALL_UNWIND_AND_CONTINUE_HANDLER_EX` /           `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER_EX` ?

janvorli replied:
Done, I was wondering what the _NO_PROBE originally meant.

janvorli committed:
Rename INSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE

To INSTALL_UNWIND_AND_CONTINUE_HANDLER_EX, as the old name is obsolete
